### PR TITLE
feat(VM): Switch from Cirrus video adaptor to VGA adaptor

### PR DESCRIPTION
**Pull request by badrAZ**
Fixes vatesfr/xo-web#158

### CODE REVIEW

File: src/xapi-object-to-xo.js [hunk 1]
```diff
@@ -1,5 +1,4 @@
 import {
-  includes,
```
Comment (badrAZ) [inline src/xapi-object-to-xo.js]: I removed it because i got a standard error.

File: src/api/vm.coffee [hunk 2]
```diff
@@ -544,6 +544,11 @@ set.params = {
   cpuCap: { type: ['integer', 'null'], optional: true }
 
   affinityHost: { type: ['string', 'null'], optional: true }
+
+  # Switch from Cirrus video adaptor to VGA adaptor
+  VGA:{ type: 'boolean', optional: true }
```
Comment (julien-f) [inline src/api/vm.coffee]: Camel case: `vga`.

File: src/xapi/mixins/vm.js [hunk 3]
```diff
@@ -341,7 +341,24 @@ export default {
 
     PV_args: true,
 
-    tags: true
+    tags: true,
+
+    VGA: {
```
Comment (julien-f) [inline src/xapi/mixins/vm.js]: Camel case.

File: src/xapi/mixins/vm.js [hunk 4]
```diff
@@ -341,7 +341,24 @@ export default {
...
+      set (vga, vm) {
+        return this._updateObjectMapProperty(vm, 'platform', {
+          'vga': vga ? 'std' : null,
+          'videoram': vga ? 4 : null
```
Comment (julien-f) [inline src/xapi/mixins/vm.js]: So, when setting `vga` you are also overriding `videoram`?

File: src/xapi/mixins/vm.js [hunk 5]
```diff
@@ -341,7 +341,24 @@ export default {
...
+        })
+      }
+    },
+
+    videoram: {
+      set (videoram, vm) {
+        // according to: https://xenserver.org/blog/entry/vga-over-cirrus-in-xenserver-6-2.html
+        // the maximal value that the video RAM can take is 16
+        return this._updateObjectMapProperty(vm, 'platform', { 'videoram': videoram > 16 ? 16 : videoram })
+      }
```
Comment (julien-f) [inline src/xapi/mixins/vm.js]: What's happening if both `vga` and `videoram` are set at the same time?

File: src/xapi/other-config-template.js [hunk 6]
```diff
@@ -26,6 +26,8 @@ const OTHER_CONFIG_TEMPLATE = {
   },
   PCI_bus: '',
   platform: {
+    videoram: '4',
+    vga: 'std',
```
Comment (julien-f) [inline src/xapi/other-config-template.js]: Why this change?

File: src/xapi/mixins/vm.js (hunk 5 above)
Comment (badrAZ) [inline src/xapi/mixins/vm.js]: In the UI, there is a toggle for vga, when it is enabled the videoram input appear.

Comment (julien-f) [inline src/xapi/mixins/vm.js]: Yes, but all methods can also be called from xo-cli.

File: src/xapi/mixins/vm.js (hunk 4 above)
Comment (badrAZ) [inline src/xapi/mixins/vm.js]: When the vga is enabled, the default value of the videoram is 4.

File: src/xapi-object-to-xo.js [hunk 7]
```diff
@@ -359,6 +359,11 @@ const TRANSFORMS = {
       $VGPUs: link(obj, 'VGPUs')
     }
 
+    if (isHvm) {
+      vm.vga = obj.platform.vga
```
Comment (julien-f) [inline src/xapi-object-to-xo.js]: `vm.vga` should be a boolean.

File: src/xapi-object-to-xo.js [hunk 8]
```diff
@@ -359,6 +359,11 @@ const TRANSFORMS = {
...
+      vm.videoram = obj.platform.videoram
```
Comment (julien-f) [inline src/xapi-object-to-xo.js]: Is it relevant when `vga` is not true?

File: src/xapi/mixins/vm.js [hunk 9]
```diff
@@ -341,7 +341,29 @@ export default {
...
+    },
+
+    videoram: {
+      set (videoram, vm) {
+        // according to: https://xenserver.org/blog/entry/vga-over-cirrus-in-xenserver-6-2.html
+        // the maximal value that the video RAM can take is 16
+        const videoramValues = [1, 2, 4, 8, 16]
+
+        if (videoramValues.indexOf(+videoram) === -1) {
```
Comment (julien-f) [inline src/xapi/mixins/vm.js]: Avoid using `Array#indexOf`, you can use an object instead.

File: src/xapi/mixins/vm.js [hunk 10]
```diff
@@ -341,7 +341,29 @@ export default {
...
+          throw new Error('The different values that the video RAM can take are: 1,2,4,8,16')
```
Comment (julien-f) [inline src/xapi/mixins/vm.js]: Avoid copying: use values in `videoramValues`.

File: src/xapi/mixins/vm.js [hunk 11]
```diff
@@ -341,7 +341,29 @@ export default {
...
+          'vga': vga === true ? 'std' : null
+        })
+      }
+    },
+
+    videoram: {
+      set (videoram, vm) {
+        // according to: https://xenserver.org/blog/entry/vga-over-cirrus-in-xenserver-6-2.html
+        // the maximal value that the video RAM can take is 16
+        const videoramValues = [1, 2, 4, 8, 16]
```
Comment (julien-f) [inline src/xapi/mixins/vm.js]: Should be defined outside of this function.

File: src/xapi/mixins/vm.js [hunk 12]
```diff
@@ -341,7 +341,29 @@ export default {
...
+
+        if (videoramValues.indexOf(+videoram) === -1) {
+          throw new Error('The different values that the video RAM can take are: 1,2,4,8,16')
+        } else {
```
Comment (julien-f) [inline src/xapi/mixins/vm.js]: `else` is unnecessary here.

File: src/xapi/mixins/vm.js [hunk 13]
```diff
@@ -341,7 +341,29 @@ export default {
...
+          return this._updateObjectMapProperty(vm, 'platform', { 'videoram': videoram })
```
Comment (julien-f) [inline src/xapi/mixins/vm.js]: Unnecessary quotes in `'videoram'`, and you can use the shorthand notation.

File: src/api/vm.coffee [hunk 14]
```diff
@@ -544,6 +544,11 @@ set.params = {
   cpuCap: { type: ['integer', 'null'], optional: true }
 
   affinityHost: { type: ['string', 'null'], optional: true }
+
+  # Switch from Cirrus video adaptor to VGA adaptor
+  vga:{ type: 'boolean', optional: true }
```
Comment (julien-f) [inline src/api/vm.coffee]: Missing space after `vga:`.

File: src/xapi/mixins/vm.js [hunk 15]
```diff
@@ -341,7 +341,29 @@ export default {
 
     PV_args: true,
 
-    tags: true
+    tags: true,
+
+    vga: {
+      set (vga, vm) {
+        return this._updateObjectMapProperty(vm, 'platform', {
+          'vga': vga === true ? 'std' : null
```
Comment (julien-f) [inline src/xapi/mixins/vm.js]: Unnecessary quotes for `'vga'`.

File: src/xapi-object-to-xo.js (hunk 8 above)
Comment (badrAZ) [inline src/xapi-object-to-xo.js]: Xen Server ignore the parameter when it switched to cirrus and it is used when we come back to vga.

File: src/xapi-object-to-xo.js [hunk 16]
```diff
@@ -360,7 +360,7 @@ const TRANSFORMS = {
     }
 
     if (isHvm) {
-      vm.vga = obj.platform.vga
+      vm.vga = (obj.platform.vga !== undefined)
```
Comment (julien-f) [inline src/xapi-object-to-xo.js]: Parentheses are unnecessary.

Shouldn't the test be `=== 'std'`?

File: src/xapi/mixins/vm.js [hunk 17]
```diff
@@ -19,6 +20,10 @@ import {
   NULL_REF
 } from '../utils'
 
+// According to: https://xenserver.org/blog/entry/vga-over-cirrus-in-xenserver-6-2.html,
+// the maximal value that the video RAM can take is 16.
+const XEN_VIDEORAM_VALUES = ['null', 1, 2, 4, 8, 16]
```
Comment (julien-f) [inline src/xapi/mixins/vm.js]: How is `'null`' a valid value?

File: src/xapi/mixins/vm.js [hunk 18]
```diff
@@ -346,22 +351,17 @@ export default {
...
       set (videoram, vm) {
-        // according to: https://xenserver.org/blog/entry/vga-over-cirrus-in-xenserver-6-2.html
-        // the maximal value that the video RAM can take is 16
-        const videoramValues = [1, 2, 4, 8, 16]
-
-        if (videoramValues.indexOf(+videoram) === -1) {
-          throw new Error('The different values that the video RAM can take are: 1,2,4,8,16')
-        } else {
-          return this._updateObjectMapProperty(vm, 'platform', { 'videoram': videoram })
+        if (some(XEN_VIDEORAM_VALUES, videoram) === false) {
```
Comment (julien-f) [inline src/xapi/mixins/vm.js]: Use `lodash/includes`.

File: src/xapi/mixins/vm.js (hunk 17 above)
Comment (badrAZ) [inline src/xapi/mixins/vm.js]: 'null' is used to remove the parameter.

File: src/xapi-object-to-xo.js (hunk 16 above)
Comment (badrAZ) [inline src/xapi-object-to-xo.js]: Yes, i will fix it.

File: src/xapi/mixins/vm.js (hunk 17 above)
Comment (julien-f) [inline src/xapi/mixins/vm.js]: No, `null` is used to remove the value, not `'null'`.

File: src/xapi/mixins/vm.js [hunk 19]
```diff
@@ -358,7 +358,7 @@ export default {
 
     videoram: {
       set (videoram, vm) {
-        if (some(XEN_VIDEORAM_VALUES, videoram) === false) {
+        if (includes(XEN_VIDEORAM_VALUES, videoram) === false) {
```
Comment (julien-f) [inline src/xapi/mixins/vm.js]: Use `!` not `=== false`.

File: src/xapi/mixins/vm.js [hunk 20]
```diff
@@ -3,7 +3,7 @@ import find from 'lodash/find'
 import gte from 'lodash/gte'
 import isEmpty from 'lodash/isEmpty'
 import lte from 'lodash/lte'
-import some from 'lodash/some'
+import includes from 'lodash/includes'
```
Comment (julien-f) [inline src/xapi/mixins/vm.js]: Keep imports sorted.

File: src/xapi/mixins/vm.js [hunk 21]
```diff
@@ -22,7 +22,7 @@ import {
 
 // According to: https://xenserver.org/blog/entry/vga-over-cirrus-in-xenserver-6-2.html,
 // the maximal value that the video RAM can take is 16.
```
Comment (julien-f) [inline src/xapi/mixins/vm.js]: This comment do not provide any new info, just keep the link.

File: src/xapi/mixins/vm.js [hunk 22]
```diff
@@ -19,6 +20,10 @@ import {
   NULL_REF
 } from '../utils'
 
+// According to: https://xenserver.org/blog/entry/vga-over-cirrus-in-xenserver-6-2.html.
+const XEN_VGA_VALUES = ['std', 'cirrus']
```
Comment (julien-f) [inline src/xapi/mixins/vm.js]: What about setting the default value?

File: src/xapi-object-to-xo.js [hunk 23]
```diff
@@ -361,7 +361,8 @@ const TRANSFORMS = {
 
     if (isHvm) {
       vm.vga = obj.platform.vga
-      vm.videoram = obj.platform.videoram
+
+      if (vm.vga === 'std') vm.videoram = obj.platform.videoram || 4
```
Comment (julien-f) [inline src/xapi-object-to-xo.js]:
```js
({
  vga: vm.vga = 'cirrus',
  videoram: vm.videoram = 4
} = obj.platform)
```

File: src/api/vm.coffee [hunk 24]
```diff
@@ -544,6 +544,11 @@ set.params = {
   cpuCap: { type: ['integer', 'null'], optional: true }
 
   affinityHost: { type: ['string', 'null'], optional: true }
+
+  # Switch from Cirrus video adaptor to VGA adaptor
+  vga: { type: 'string', optional: true }
+
+  videoram: { type: ['string', 'number', 'null'], optional: true }
```
Comment (julien-f) [inline src/api/vm.coffee]: `null` is no longer allowed.
